File: alljoyn/services/time/cpp/src/client/TimeServiceClient.cc
/******************************************************************************
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

#include <alljoyn/time/TimeServiceClient.h>
#include <alljoyn/time/LogModule.h>
#include <alljoyn/time/TimeServiceConstants.h>
#include <alljoyn/time/TimeServiceSessionListener.h>
#include "TimeServiceSessionHandler.h"

using namespace ajn;
using namespace services;

//Constructor
TimeServiceClient::TimeServiceClient() : m_Bus(NULL), m_SessionHandler(NULL)
{

    QCC_DbgTrace(("%s", __FUNCTION__));
}

//Destructor
TimeServiceClient::~TimeServiceClient()
{

    QCC_DbgTrace(("%s", __FUNCTION__));
    release();
}


//Initialize the object
QStatus TimeServiceClient::init(BusAttachment* bus, const qcc::String& serverBusName, const qcc::String& deviceId,
                                const qcc::String& appId, const ajn::AboutObjectDescription& aboutObjectDescription)
{

    QCC_DbgTrace(("%s, ServerBusName: '%s', DeviceId: '%s', AppId: '%s'", __FUNCTION__, serverBusName.c_str(),
                  deviceId.c_str(), appId.c_str()));

    QStatus status = checkObjectsValidity(bus, serverBusName);
    if (status != ER_OK) {

        return status;
    }

    m_Bus               = bus;
    m_ServerBusName     = serverBusName;
    m_DeviceId          = deviceId;
    m_AppId             = appId;

    QCC_DbgHLPrintf(("Initializing TimeServiceClient for the server: '%s'", m_ServerBusName.c_str()));

    initManagerMap();
    analyzeObjectDescriptions(aboutObjectDescription);

    return ER_OK;
}

//Release object resources
void TimeServiceClient::release()
{

    QCC_DbgTrace(("%s, ServerBusName: '%s', DeviceId: '%s', AppId: '%s'", __FUNCTION__, m_ServerBusName.c_str(),
                  m_DeviceId.c_str(), m_AppId.c_str()));

    QCC_DbgHLPrintf(("Releasing TimeServiceClient for the server: '%s'", m_ServerBusName.c_str()));

    //When TSItemsManager object is destructed, it iteratively deletes all its TimeService objects
    for (std::map<qcc::String, TsItemsManagerBase*>::iterator iter = m_ManagerMap.begin(); iter != m_ManagerMap.end(); ++iter) {

        delete iter->second;
    }

    m_ManagerMap.clear();

    if (m_SessionHandler) {

        delete m_SessionHandler;
        m_SessionHandler = NULL;
    }

    m_Bus = NULL;
}

//Return BusAttachment
BusAttachment* TimeServiceClient::getBus() const
{

    return m_Bus;
}

//The bus name that is used to reach TimeServiceServer
const qcc::String& TimeServiceClient::getServerBusName() const
{

    return m_ServerBusName;
}

//The id of the device hosting the TimeServiceServer
const qcc::String& TimeServiceClient::getDeviceId() const
{

    return m_DeviceId;
}

//The name of the application that started the TimeServiceServer
const qcc::String& TimeServiceClient::getAppId() const
{

    return m_AppId;
}

//Returns list of the TimeServiceClientClock objects found in the received object descriptions
const std::vector<TimeServiceClientClock*>& TimeServiceClient::getAnnouncedClockList() const
{

    return m_Clocks;
}

//Returns list of the TimeServiceClientAlarm objects found in the received object descriptions
const std::vector<TimeServiceClientAlarm*>& TimeServiceClient::getAnnouncedAlarmList() const
{

    return m_Alarms;
}

//Returns list of the TimeServiceClientAlarmFactory objects found in the received object descriptions
const std::vector<TimeServiceClientAlarmFactory*>& TimeServiceClient::getAnnouncedAlarmFactoryList() const
{

    return m_AlarmFactories;
}

//Returns list of the TimeServiceClientTimer objects found in the received object descriptions
const std::vector<TimeServiceClientTimer*>& TimeServiceClient::getAnnouncedTimerList() const
{

    return m_Timers;
}

//Returns list of the TimeServiceClientTimerFactory objects found in the received object descriptions
const std::vector<TimeServiceClientTimerFactory*>& TimeServiceClient::getAnnouncedTimerFactoryList() const
{

    return m_TimerFactories;
}

//Creates the session asynchronously with the server.
QStatus TimeServiceClient::joinSessionAsync(TimeServiceSessionListener* sessionListener)
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    if (!m_SessionHandler) {

        m_SessionHandler = new TimeServiceSessionHandler(this);
    }

    return m_SessionHandler->joinSessionAsync(sessionListener);
}

//Leave previously established session
QStatus TimeServiceClient::leaveSession()
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    if (!m_SessionHandler) {

        QCC_LogError(ER_FAIL, ("The session hasn't been joined before"));
        return ER_FAIL;
    }

    return m_SessionHandler->leaveSession();
}

//Returns TRUE if there is a session established with the server
bool TimeServiceClient::isConnected() const
{

    if (!m_SessionHandler) {

        QCC_LogError(ER_FAIL, ("The session hasn't been joined before"));
        return false;
    }

    return m_SessionHandler->isConnected();
}

//Returns current session id, use the method isConnected()
QStatus TimeServiceClient::getSessionId(ajn::SessionId& sessionId) const
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    if (!m_SessionHandler) {

        QCC_LogError(ER_FAIL, ("The session hasn't been joined before"));
        return ER_FAIL;
    }

    if (!m_SessionHandler->isConnected()) {

        QCC_LogError(ER_FAIL, ("The session hasn't been established"));
        return ER_FAIL;
    }

    sessionId = m_SessionHandler->getSessionId();

    return ER_OK;
}

//Checks validity of the received BusAttachment and serverBusName
QStatus TimeServiceClient::checkObjectsValidity(BusAttachment* bus, const qcc::String& serverBusName)
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    if (m_Bus != NULL) {

        QCC_LogError(ER_FAIL, ("Could not accept this BusAttachment, a different bus attachment already exists"));
        return ER_BAD_ARG_1;
    }

    if (!bus || !bus->IsStarted() || !bus->IsConnected()) {

        QCC_LogError(ER_BAD_ARG_1, ("Could not accept this BusAttachment, bus attachment not started or not connected"));
        return ER_BAD_ARG_1;
    }

    if (serverBusName.length() == 0) {

        QCC_LogError(ER_BAD_ARG_2, ("serverBusName is empty"));
        return ER_BAD_ARG_2;
    }

    return ER_OK;
}

//Populate ManagerMap with the relevant TimeService interface names
void TimeServiceClient::initManagerMap()
{

    m_ManagerMap.insert(std::pair<qcc::String, TsItemsManagerBase*>(tsConsts::ALARM_IFACE,
                                                                    new TSItemsManager<TimeServiceClientAlarm>(*this, &m_Alarms)));

    m_ManagerMap.insert(std::pair<qcc::String, TsItemsManagerBase*>(tsConsts::ALARM_FACTORY_IFACE,
                                                                    new TSItemsManager<TimeServiceClientAlarmFactory>(*this, &m_AlarmFactories)));

    m_ManagerMap.insert(std::pair<qcc::String, TsItemsManagerBase*>(tsConsts::CLOCK_IFACE,
                                                                    new TSItemsManager<TimeServiceClientClock>(*this, &m_Clocks)));

    m_ManagerMap.insert(std::pair<qcc::String, TsItemsManagerBase*>(tsConsts::TIMER_IFACE,
                                                                    new TSItemsManager<TimeServiceClientTimer>(*this, &m_Timers)));

    m_ManagerMap.insert(std::pair<qcc::String, TsItemsManagerBase*>(tsConsts::TIMER_FACTORY_IFACE,
                                                                    new TSItemsManager<TimeServiceClientTimerFactory>(*this, &m_TimerFactories)));
}

//Analyze received object descriptions and create relevant TimeService objects
void TimeServiceClient::analyzeObjectDescriptions(const AboutObjectDescription& aboutObjectDescription)
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    size_t numPaths = aboutObjectDescription.GetPaths(NULL, 0);
    const char** paths = new const char*[numPaths];
    aboutObjectDescription.GetPaths(paths, numPaths);
    for (size_t p = 0; p < numPaths; p++) {
        qcc::String objectPath(paths[p]);
        size_t numInterfaces = aboutObjectDescription.GetInterfaces(objectPath.c_str(), NULL, 0);
        const char** interfaces = new const char*[numInterfaces];
        aboutObjectDescription.GetInterfaces(objectPath.c_str(), interfaces, numInterfaces);
        bool isAuthority = false;
        TimeServiceClientClock* currentClock = NULL;

        for (size_t i = 0; i < numInterfaces; i++) {
            qcc::String iface(interfaces[i]);

            //If IFNAME_PREFIX is not a prefix of the iface, then It's not a TimeService interface
            if (iface.find(tsConsts::IFNAME_PREFIX) == qcc::String::npos) {

                continue;
            }

            //Check if Interface is a TimeAuthority
            if (iface.compare(tsConsts::TIME_AUTHORITY_IFACE) == 0) {

                isAuthority = true;
                continue;
            }

            //Check if current interface is a known TimeService interface, if TRUE create the object using the TSItemsManager
            TSItemsManagerMap::iterator mgrIter = m_ManagerMap.find(iface.c_str());

            if (mgrIter == m_ManagerMap.end()) {

                QCC_DbgPrintf(("Received unsupported TimeService interface : '%s', objectPath: '%s', from the server: '%s'", iface.c_str(),
                               objectPath.c_str(), m_ServerBusName.c_str()));
                continue;
            }

            QCC_DbgPrintf(("Creating TimeClient object: '%s', interface: '%s', from the server: '%s'", objectPath.c_str(),
                           iface.c_str(), m_ServerBusName.c_str()));

            //Check if current interface is Clock, create Clock object and save it
            if (iface.compare(tsConsts::CLOCK_IFACE) == 0) {

                currentClock = static_cast<TimeServiceClientClock*>(mgrIter->second->create(objectPath));
            } else {

                mgrIter->second->create(objectPath);
            }

        }        //for::ifaces
        delete [] interfaces;

        if (isAuthority && currentClock) {

            QCC_DbgPrintf(("TimeClient Clock object: '%s', IsAuthority: '%d'", objectPath.c_str(), isAuthority));
            currentClock->setAuthority(isAuthority);
        }
    }    //for p
    delete [] paths;
}

//Base class constructor for the TsItemsManager providing its API
TimeServiceClient::TsItemsManagerBase::TsItemsManagerBase()
{

    QCC_DbgTrace(("%s", __FUNCTION__));
}

//Base class destructor for the TsItemsManager providing its API
TimeServiceClient::TsItemsManagerBase::~TsItemsManagerBase()
{

    QCC_DbgTrace(("%s", __FUNCTION__));
}


//TSItemsManager Constructor
template <class TSItem>
TimeServiceClient::TSItemsManager<TSItem>::TSItemsManager(const TimeServiceClient& tsClient, std::vector<TSItem*>* storage) :
    TsItemsManagerBase(),
    m_Storage(storage),
    m_TsClient(tsClient)
{

    QCC_DbgTrace(("%s", __FUNCTION__));
}


//TSItemsManager destructor Constructor
template <class TSItem>
TimeServiceClient::TSItemsManager<TSItem>::~TSItemsManager()
{

    QCC_DbgTrace(("%s", __FUNCTION__));
    release();
}

//Create TimeService item
template <class TSItem>
TimeServiceClientBase* TimeServiceClient::TSItemsManager<TSItem>::create(const qcc::String& objectPath)
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    TSItem* tsItem = new TSItem(m_TsClient, objectPath);
    m_Storage->push_back(tsItem);

    return tsItem;
}

//Release TimeService objects stored in the storage
template <class TSItem>
void TimeServiceClient::TSItemsManager<TSItem>::release()
{

    QCC_DbgTrace(("%s", __FUNCTION__));

    //After deleting the TSItem object its destructor calls release() method on the object
    for (typename std::vector<TSItem*>::iterator iter = m_Storage->begin(); iter != m_Storage->end(); ++iter) {

        delete *iter;
    }

    m_Storage->clear();
}

File: alljoyn/services/time/java/TimeService/src/org/allseen/timeservice/client/TimeServiceClient.java
/******************************************************************************
 * Copyright AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 ******************************************************************************/

package org.allseen.timeservice.client;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.alljoyn.bus.AboutObjectDescription;
import org.alljoyn.bus.BusAttachment;
import org.alljoyn.bus.Status;
import org.alljoyn.services.common.AnnouncementHandler;
import org.alljoyn.services.common.BusObjectDescription;
import org.allseen.timeservice.TimeServiceConst;
import org.allseen.timeservice.ajinterfaces.TimeAuthority;
import org.allseen.timeservice.server.TimeServiceServer;

import android.util.Log;

/**
 * The main class to initialize the Time Service in the client mode. The class
 * is initialized with the {@link BusObjectDescription} array received from the
 * {@link AnnouncementHandler#onAnnouncement(String, short, BusObjectDescription[], java.util.Map)}
 * . To retrieve the Time Service items that were found in the provided
 * {@link BusObjectDescription} array, use getter methods such as
 * {@link TimeServiceClient#getAnnouncedClockList()}.
 */
public class TimeServiceClient {
    private static final String TAG = "ajts" + TimeServiceClient.class.getSimpleName();

    /**
     * Helper class for creation or releasing Time Service objects following
     * analyzing of the received Announcement signal.
     * 
     * @param <T>
     *            The type of the created object
     */
    private class TSItemsManager<T extends TimeClientBase> {

        /**
         * The {@link Class} object of the created Time Service object
         */
        private final Class<T> createdClass;

        /**
         * List in the {@link TimeServiceClient} where to store the created
         * object
         */
        private List<T> storage;

        /**
         * Constructor
         * 
         * @param createdClass
         *            The {@link Class} object of the created Time Service
         *            object
         * @param storage
         *            List in the {@link TimeServiceClient} where to store the
         *            created object
         */
        TSItemsManager(Class<T> createdClass, List<T> storage) {

            this.createdClass = createdClass;
            this.storage = storage;
        }

        /**
         * Create the object with the given objPath
         * 
         * @param objectPath
         *            objectPath to create the Time Service item object
         * @return Created object
         * @throws Exception
         *             Is thrown if failed to create the object
         */
        public Object create(String objectPath) throws Exception {

            T createdObj = createdClass.getConstructor(TimeServiceClient.class, String.class).newInstance(TimeServiceClient.this, objectPath);

            storage.add(createdObj);
            return createdObj;
        }

        /**
         * Releases all the objects in the storage and cleans the storage.
         */
        public void release() {

            for (T tObj : storage) {

                tObj.release();
            }

            storage.clear();
            storage = null;
        }
    }

    // ==========================================================//

    /**
     * The {@link BusAttachment} that is used by the {@link TimeServiceClient}
     * to reach the {@link TimeServiceServer}
     */
    private BusAttachment bus;

    /**
     * The bus name to reach the {@link TimeServiceServer}
     */
    private final String serverBusName;

    /**
     * The id if the device hosting the {@link TimeServiceServer}
     */
    private final String deviceId;

    /**
     * The name of the application that started the {@link TimeServiceServer}
     */
    private final UUID appId;

    /**
     * Manages session with the {@link TimeServiceServer}
     */
    private TimeServiceSessionListener sessionListener;

    /**
     * Map from the announced AJ interface to the {@link TSItemsManager} that
     * allows to create and release Time Service objects
     */
    private Map<String, TSItemsManager<?>> managerMap;

    /**
     * The list of {@link Clock} objects that were found in the Announcement
     */
    private final List<Clock> clocks;

    /**
     * The list of {@link Alarm} objects that were found in the Announcement
     */
    private final List<Alarm> alarms;

    /**
     * The list of {@link Timer} objects that were found in the Announcement
     */
    private final List<Timer> timers;

    /**
     * The list of {@link AlarmFactory} objects that were found in the
     * Announcement
     */
    private final List<AlarmFactory> alarmFactories;

    /**
     * The list of {@link TimerFactory} objects that were found in the
     * Announcement
     */
    private final List<TimerFactory> timerFactories;

    /**
     * Constructor
     * 
     * @param bus
     *            The {@link BusAttachment} that is used by the
     *            {@link TimeServiceClient} to reach the
     *            {@link TimeServiceServer}
     * @param serverBusName
     *            The bus name to reach the {@link TimeServiceServer}
     * @param deviceId
     *            The id if the device hosting the {@link TimeServiceServer}
     * @param appId
     *            The name of the application that started the
     *            {@link TimeServiceServer}
     * @param objDescs
     *            Array of {@link BusObjectDescription} received with
     *            Announcement
     * @throws IllegalArgumentException
     *             Is thrown if the given {@link BusAttachment} is not
     *             initialized or not connected to the Router.
     */
    public TimeServiceClient(BusAttachment bus, String serverBusName, String deviceId, UUID appId, AboutObjectDescription[] objDescs) throws IllegalArgumentException {

        checkObjectValidity(bus, serverBusName);

        this.bus = bus;
        this.serverBusName = serverBusName;
        this.deviceId = deviceId;
        this.appId = appId;

        clocks = new ArrayList<Clock>();
        alarms = new ArrayList<Alarm>();
        timers = new ArrayList<Timer>();
        alarmFactories = new ArrayList<AlarmFactory>();
        timerFactories = new ArrayList<TimerFactory>();

        managerMap = new HashMap<String, TSItemsManager<?>>();
        initFactoryMap();
        analyzeObjectDescriptions(objDescs);
    }

    /**
     * The {@link BusAttachment} that is used by the {@link TimeServiceClient}
     * to reach the {@link TimeServiceServer}
     * 
     * @return {@link BusAttachment}
     */
    public BusAttachment getBus() {
        return bus;
    }

    /**
     * The bus name to reach the {@link TimeServiceServer}
     * 
     * @return The bus name of the Time Service Server
     */
    public String getServerBusName() {
        return serverBusName;
    }

    /**
     * The id if the device hosting the {@link TimeServiceServer}
     * 
     * @return device Id
     */
    public String getDeviceId() {
        return deviceId;
    }

    /**
     * The name of the application that started the {@link TimeServiceServer}
     * 
     * @return The application id
     */
    public UUID getAppId() {
        return appId;
    }

    /**
     * List of the {@link Clock} objects found in the received object
     * descriptions
     * 
     * @return List of {@link Clock} objects
     */
    public List<Clock> getAnnouncedClockList() {

        return Collections.unmodifiableList(clocks);
    }

    /**
     * List of the {@link Alarm} objects found in the received object
     * descriptions
     * 
     * @return List of {@link Alarm} objects
     */
    public List<Alarm> getAnnouncedAlarmList() {

        return Collections.unmodifiableList(alarms);
    }

    /**
     * List of the {@link AlarmFactory} objects found in the received object
     * descriptions
     * 
     * @return List of {@link AlarmFactory} objects
     */
    public List<AlarmFactory> getAnnouncedAlarmFactoryList() {

        return Collections.unmodifiableList(alarmFactories);
    }

    /**
     * List of the {@link Timer} objects found in the received object
     * descriptions
     * 
     * @return List of {@link Timer} objects
     */
    public List<Timer> getAnnouncedTimerList() {

        return Collections.unmodifiableList(timers);
    }

    /**
     * List of the {@link TimerFactory} objects found in the received object
     * descriptions
     * 
     * @return List of {@link TimerFactory} objects
     */
    public List<TimerFactory> getAnnouncedTimerFactoryList() {

        return Collections.unmodifiableList(timerFactories);
    }

    /**
     * Returns TRUE if {@link org.allseen.timeservice.ajinterfaces.Clock}
     * interface was found in the received {@link BusObjectDescription} array.
     * If this method returns TRUE,
     * {@link TimeServiceClient#getAnnouncedClockList()} will return a
     * {@link List} of found {@link Clock}s.
     * 
     * @return TRUE if {@link Clock} objects were found in the received
     *         {@link BusObjectDescription} array.
     */
    public boolean isClockAnnounced() {

        return clocks.size() != 0;
    }

    /**
     * Returns TRUE if {@link org.allseen.timeservice.ajinterfaces.Alarm}
     * interface was found in the received {@link BusObjectDescription} array.
     * If this method returns TRUE,
     * {@link TimeServiceClient#getAnnouncedAlarmList()} will return a
     * {@link List} of found {@link Alarm}s.
     * 
     * @return TRUE if {@link Alarm} objects were found in the received
     *         {@link BusObjectDescription} array.
     */
    public boolean isAlarmAnnounced() {

        return alarms.size() != 0;
    }

    /**
     * Returns TRUE if {@link org.allseen.timeservice.ajinterfaces.AlarmFactory}
     * interface was found in the received {@link BusObjectDescription} array.
     * If this method returns TRUE,
     * {@link TimeServiceClient#getAnnouncedAlarmFactoryList()} will return a
     * {@link List} of found {@link AlarmFactory}s.
     * 
     * @return TRUE if {@link AlarmFactory} objects were found in the received
     *         {@link BusObjectDescription} array.
     */
    public boolean isAlarmFactoryAnnounced() {

        return alarmFactories.size() != 0;
    }

    /**
     * Returns TRUE if {@link org.allseen.timeservice.ajinterfaces.Timer}
     * interface was found in the received {@link BusObjectDescription} array.
     * If this method returns TRUE,
     * {@link TimeServiceClient#getAnnouncedTimerList()} will return a
     * {@link List} of found {@link Timer}s.
     * 
     * @return TRUE if {@link Timer} objects were found in the received
     *         {@link BusObjectDescription} array.
     */
    public boolean isTimerAnnounced() {

        return timers.size() != 0;
    }

    /**
     * Returns TRUE if {@link org.allseen.timeservice.ajinterfaces.TimerFactory}
     * interface was found in the received {@link BusObjectDescription} array.
     * If this method returns TRUE,
     * {@link TimeServiceClient#getAnnouncedTimerFactoryList()} will return a
     * {@link List} of found {@link TimerFactory}s.
     * 
     * @return TRUE if {@link TimerFactory} objects were found in the received
     *         {@link BusObjectDescription} array.
     */
    public boolean isTimerFactoryAnnounced() {

        return timerFactories.size() != 0;
    }

    /**
     * Return whether a session with the {@link TimeServiceServer} bus has been
     * created
     * 
     * @return TRUE if the session has been created
     */
    public boolean isConnected() {

        return getSessionId() != null;
    }

    /**
     * Returns current session id or NULL if the session is not established or
     * lost.
     * 
     * @return session id
     */
    public Integer getSessionId() {

        if (sessionListener == null) {

            return null;
        }

        return sessionListener.getSessionId();
    }

    /**
     * Cleans the object and releases its resources <br>
     * It's a programming error to call another method on this object after the
     * release method has been called.
     */
    public void release() {

        Log.i(TAG, "Release was called, cleaning TimeService objects");

        // Prevent usage of the TimeServiceClient stale object
        if (managerMap != null) {

            for (TSItemsManager<?> manager : managerMap.values()) {

                manager.release();
            }

            managerMap.clear();
            managerMap = null;
        }

        if (sessionListener != null) {

            sessionListener.release();
        }

        // Last command for all the users of this BusAttachment for the release
        // purposes
        bus = null;
    }

    /**
     * Creates the session asynchronously with the {@link TimeServiceServer}
     * bus.
     * 
     * @param sessionListenerHandler
     *            Session related events will be passed via the
     *            {@link SessionListenerHandler}
     */
    public void joinSessionAsync(SessionListenerHandler sessionListenerHandler) {

        if (sessionListener == null) {

            sessionListener = new TimeServiceSessionListener(this);
        }

        sessionListener.joinSessionAsync(sessionListenerHandler);
    }

    /**
     * Leave previously established session
     * 
     * @return {@link Status} the method call status
     * @throws IllegalStateException
     *             if
     *             {@link TimeServiceClient#joinSessionAsync(SessionListenerHandler)}
     *             method has never called before
     * @see BusAttachment#leaveSession(int)
     * @see Status
     */
    public Status leaveSession() {

        if (sessionListener == null) {

            throw new IllegalStateException("Join session method has never called before");
        }

        return sessionListener.leaveSession();
    }

    /**
     * Checks validity of the received arguments. Objects should be initialized.
     * {@link BusAttachment} should be connected to the Router.
     * 
     * @param bus
     *            {@link BusAttachment} to test, should be initialized and
     *            connected.
     * @param serverBusName
     *            The bus name of the {@link TimeServiceServer}.
     * @throws IllegalArgumentException
     *             if non valid arguments were received
     */
    private void checkObjectValidity(BusAttachment bus, String serverBusName) {

        if (bus == null) {

            throw new IllegalArgumentException("BusAttachment is not initialized");
        }

        if (!bus.isConnected()) {

            throw new IllegalArgumentException("BusAttachment is not connected");
        }

        if (serverBusName == null || serverBusName.length() == 0) {

            throw new IllegalArgumentException("serverBusName is not initialized");
        }
    }

    /**
     * Initialize the managerMao with the required for object creation data
     */
    private void initFactoryMap() {

        managerMap.put(org.allseen.timeservice.ajinterfaces.Alarm.IFNAME, new TSItemsManager<Alarm>(Alarm.class, alarms));
        managerMap.put(org.allseen.timeservice.ajinterfaces.Clock.IFNAME, new TSItemsManager<Clock>(Clock.class, clocks));
        managerMap.put(org.allseen.timeservice.ajinterfaces.Timer.IFNAME, new TSItemsManager<Timer>(Timer.class, timers));
        managerMap.put(org.allseen.timeservice.ajinterfaces.AlarmFactory.IFNAME, new TSItemsManager<AlarmFactory>(AlarmFactory.class, alarmFactories));

        managerMap.put(org.allseen.timeservice.ajinterfaces.TimerFactory.IFNAME, new TSItemsManager<TimerFactory>(TimerFactory.class, timerFactories));
    }

    /**
     * Iterates over the array of {@link BusObjectDescription} and creates the
     * Time Service objects in depend on the found interfaces
     * 
     * @param objectDescs
     */
    private void analyzeObjectDescriptions(AboutObjectDescription[] objectDescs) {

        for (AboutObjectDescription bod : objectDescs) {

            String objPath = bod.path;
            String[] ifaces = bod.interfaces;

            boolean timeAuth = false; // whether current objPath implements
                                      // TimeAuthority
            Clock currClock = null;

            for (String iface : ifaces) {

                // Not a Time Service interface
                if (!iface.startsWith(TimeServiceConst.IFNAME_PREFIX)) {
                    continue;
                }

                // If the interface is TimeAuthority, store it
                if (iface.equals(TimeAuthority.IFNAME)) {

                    timeAuth = true;
                    continue;
                }

                TSItemsManager<?> manager = managerMap.get(iface);
                if (manager == null) {

                    Log.w(TAG, "Received unsupported interface: '" + iface + "', objPath: '" + objPath + "', from busName: '" + serverBusName + "'");
                    continue;
                }

                Object object;
                try {

                    object = manager.create(objPath);
                } catch (Exception e) {

                    Log.e(TAG, "Failed to create the TimeClient object, objPath: '" + objPath + "', interface: '" + iface + "' " + "received from bus: '" + serverBusName + "'", e);
                    continue;
                }

                Log.d(TAG, "Created TimeClient object, received from busName: '" + serverBusName + "' - " + object);

                if (object instanceof Clock) {

                    currClock = (Clock) object;
                }
            }// for::ifaces

            if (currClock != null && timeAuth) {

                currClock.setAuthority(timeAuth);
                Log.d(TAG, "Setting TimeAuthority for the Clock - " + currClock);
            }
        }
    }

}
